nixos/security.gnupg: provisioning GnuPG-protected secrets through the Nix store

ju1m opened the pull request:
###### Motivation for this change
This `security.gnupg` proposal aims at granularly sending, decrypting, and installing secrets (and if need be reloading services depending on them) coming from GnuPG encrypted files put into the Nix store, by leveraging `password-store`, `gpg-agent`, `systemd` and `nix copy`.

Thanks in advance for your feedbacks.

###### Features
- [X] Read from a local `password-store`'s `$PASSWORD_STORE_DIR` or any `.gpg` loadable into the Nix store by any mean available.
- [X] Check at build-time that needed secrets exist (in `config.security.pass.secrets`).
- [X] No need to be able to decrypt the secrets at build-time nor locally.
- [X] Send only secrets actually used (not the whole `password-store`), and do not resend them unless they've changed.
- [X] Load only in RAM the passphrase to use the decrypting OpenPGP key(s) (with `gpg-preset-passphrase`).
- [X] Remove volatile secrets when unconfigured (leveraging `RuntimeDirectory`), or persistent secrets only if their `postStop` is configured to do so.
- [X] Restart or reload only needed services (one file in the Nix store and one `secret-${secret}.service` per entry in `config.security.gnupg.secrets` allow fine grained `postStart`, `after`, `wants`, or `requires`).
- [X] Allow decorating secrets with non-secret bits (using the option `pipe`).
- [X] Simpler setup than using `extraBuiltins.exec` calling `pass` and/or `nixops send-keys`.
- [X] AFAIU, trusting or signing the activated NixOS generation removes the need to verify secrets.
- [X] Use a dedicated `gnupg` user instead of `root` to run `gpg-agent`.
- [X] Provide a helper (`config.security.gnupg.agent.sendKeys`) to send through `ssh` a decrypting key and its password to `gpg-agent`.

###### TODO
1. [ ] Get more real-life tests and feedbacks.
2. [ ] Put the present documentation in the manual.
3. [ ] Write a unit test in `nixos/tests/`.
4. [ ] Test using smartcard on the target machine.

###### Inconvenients
1. An OpenPGP key (preferably dedicated for the task and maybe per target machine) must be generated (with the encryption capability).
2. Encrypted secrets are readable by all users in the Nix store, if this is a concern this can maybe be restricted by leveraging `apparmor` (for which I happen to be proposing the PR https://github.com/NixOS/nixpkgs/pull/93457 ).

###### Example
First just rebuild your configuration without configuring secrets, just to enable `gpg-agent.service`, using:
```
security.gnupg.agent.enable = true;
```

Then make sure secrets are (re)encrypted to the right OpenPGP recipients (eg. `root@example.org`'s key):
```
pass init -p machines/example @admin1@ @admin2@ @root@example.org
```

Now configure and use some secret (eg. `pass hosts/foo/transmission/settings.json`) :
```
{ config, lib, pkgs, ... }:
let inherit (config.security.gnupg) secrets; in
{
# Although it is a path, it is used by `security.gnupg`
# so that only needed `.gpg` files within it will go into the Nix store
# (unless it is set to a path beginning by an entry of the `flake.nix`'s `inputs`,
# in which case it is already inside the Nix store).
security.gnupg.store = .password-store/hosts/example;

# root:root ownership with 400 mode by default
security.gnupg.secrets."transmission/settings.json" = {};

# Enable gpg-agent.service and the security.gnupg.agent.sendKeys script
# to send the encrypt subkey specified by its keygrip
# (obtained with: gpg --list-secret-keys --with-keygrip --with-keygrip)
security.gnupg.agent = {
  # Activate `gpg-agent` on the target host.
  # Unless you choose to [forward the agent](https://wiki.gnupg.org/AgentForwarding)
  # of the orchestrating host, to `/var/lib/gnupg/.gnupg/S.gpg-agent`.
  enable = true;
  keyring."9AA84E6F6D71F9163C46BF396B141A0806219077" = {
    # Defaults to config.security.gnupg.store + "/keygrip/9AA84E6F6D71F9163C46BF396B141A0806219077.gpg"
    passwordGpg = .password-store/hosts/example/root.pass.gpg;
  };
};

# It's not needed here because `credentialsFile` is set up by the `root` user
# but for some services their user would have to be added to the `keys` group.
#users.users.transmission.extraGroups = [ config.users.groups.keys.name ];

systemd.services.transmission = {
  after = [ secrets."transmission/settings.json".service ];
  # No reload for transmission: always restart if the secret has changed.
  # Use `wants` and `security.gnupg.secrets."path/to/secret".systemdConfig.postStart`
  # for a service that can be reloaded.
  requires = [ secrets."transmission/settings.json".service ];
};

services.transmission = {
  enable = true;
  credentialsFile = secrets."transmission/settings.json".path;
};
}
```

###### Without a `flake.nix`
From there you can use `<nixpkgs/nixos>` as usual, interleaving a call to `sendKeys`:
```
nix-build ./nixos.nix -A system -o nixos
nix copy --to ssh://root@example.org --substitute-on-destination ./nixos
nix-build ./nixos.nix -A security.gnupg.agent.sendKeys -o sendKeys && ./sendKeys/bin/gnupg-agent-sendKeys
ssh root@example.org "
 nix-env --profile /nix/var/nix/profiles/system --set $(readlink -m ./nixos) &&
 /nix/var/nix/profiles/system/bin/switch-to-configuration switch"
```

Logged in as `root@example.org`, you can then list the systemd units responsible to install those secrets with:
```
systemctl list-units 'secret-*' --all
```
and check the ones decrypted on the default destination, like so:
```
ls -l /run/keys/gnupg/**/file
```
###### With a `flake.nix`

```
{
inputs.nixpkgs.url = "flake:nixpkgs";
inputs.flake-utils.url = "github:numtide/flake-utils";
inputs.pass = { type = "path"; path = "./pass"; flake = false; };
outputs = inputs:
  {
    # Usage: nix -L build .#nixosConfigurations.$hostName.config.system.build.toplevel
    nixosConfigurations = builtins.mapAttrs (hostName: hostConfig:
      let cfg = import hostConfig { inherit inputs; }; in
      import (inputs.nixpkgs + "/nixos/lib/eval-config.nix") (cfg // {
        extraArgs = {
          inherit hostName inputs;
          hosts = inputs.self.nixosConfigurations;
        } // (cfg.extraArgs or {});
        modules = cfg.modules ++ [({pkgs, ...}: {
          nix.extraOptions = "experimental-features = nix-command flakes";
          security.gnupg.store = inputs.pass + "/hosts/${hostName}";
        })];
      }))
      {
        losurdo = hosts/losurdo.nix;
        mermet  = hosts/mermet.nix;
      };
  }
  // inputs.flake-utils.lib.eachDefaultSystem (system:
    let pkgs = inputs.nixpkgs.legacyPackages.${system}; in
    {
    apps = builtins.mapAttrs (hostName: { config, ... }: let
      system = config.system.build.toplevel;
      target = "root@${config.networking.hostName}.${config.networking.domain}";
      profile = "/nix/var/nix/profiles/system";
      in rec {

      # Usage: nix run .#$hostName.switch
      "switch" = {
        type = "app";
        program = (pkgs.writeShellScript "switch" ''
          set -eux
          nix-store --add-root hosts/${hostName}.root --indirect --realise ${system}
          nix copy --to ssh://'${target}' --substitute-on-destination ${system}

          ${sendkeys.program}

          ssh '${target}' \
            nix-env --profile '${profile}' --set '${system}' '&&' \
            '${profile}'/bin/switch-to-configuration switch
        '').outPath;
      };

      # Usage: nix run .#$hostName.sendkeys
      "sendkeys" = {
        type = "app";
        program = config.security.gnupg.agent.sendKeys + "/bin/gnupg-agent-sendKeys";
      };
    }) inputs.self.nixosConfigurations;}
  );
}
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
```

Mic92 commented:
Interesting I was working on this: https://github.com/Mic92/sops-nix

Mic92 commented on nixos/modules/security/pass.nix:
```diff
@@ -0,0 +1,200 @@
+             --decrypt '${secret.gpg}' \
+            ${lib.optionalString (secret.pipe != null) (" | "+secret.pipe)}
+            } |
+            install -D -m '${secret.mode}' -o '${secret.user}' -g '${secret.group}' /dev/stdin \
```
What do you think about installing it to `/run/keys` by default. In sops-nix I do this and create symlinks from their. This way secrets are kept in RAM. Combined with hardware tokens for GPG no secret would be persistently stored.

ju1m replied:
By default secrets are already installed in `/run`, and now the passphrase no longer comes from a file but only from `gpg-agent`. This passphrase can be uploaded each time with `gpg-preset-passphrase`, or made persistent in a file loaded with `gpg-preset-passphrase` in the `postStart` of `gpg-agent.service`.

Mic92 replied:
I mean `/run/keys` not `/run/` because tmpfs can be swapped to disk. But we also discussed this here: https://github.com/NixOS/nixpkgs/pull/93659#discussion_r459849419

Mic92 commented on nixos/modules/security/pass.nix:
```diff
@@ -0,0 +1,200 @@
+          set -eux
+          decrypt() {
+            {
+            ${pkgs.gnupg}/bin/gpg --batch --pinentry-mode loopback \
```
Does this not create a gpg daemon? Should be stopped afterwards or do we do not bother? Or does systemd kills it when the service finishes?

ju1m replied:
Oh! indeed... and this gpg-agent would be destroyed by `pam_systemd` removing `/run/user/0/` when the last `root` logouts. This has given me the idea to leverage `gpg-agent` to avoid writting the passphrase on disk (though it's still possible if one wants a persistent setup). For that I had to rewrite most of the module, and also decided to rename it `security.gnupg` which is a clearer name wrt. what it does now. It was a bit hard to find a clean way to bypass the wipe out of `pam_systemd` to preserve `gpg-agent` but I may have found one, as explained in the comments. Thanks for bringing this to my attention.

Mic92 commented on nixos/modules/security/pass.nix:
```diff
@@ -0,0 +1,200 @@
+        path = lib.mkOption {
+          type = types.str;
+          default = name;
+          apply = p: if match "^/.*" p == null then "/run/pass-secrets/"+p+"/file" else p;
```
Maybe put this on ramfs.

ju1m replied:
Isn't it already the case? on my NixOS I have:
```
tmpfs on /run type tmpfs (rw,nosuid,nodev,size=1990820k,mode=755)
```

ju1m replied:
oh, right, `/run/keys` really is `ramfs`, whereas `tmpfs` may put files on the swap. I'll fix that tomorrow. Thanks!

Mic92 commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,253 @@
+options.security.gnupg = {
+  store = lib.mkOption {
+    type = types.path;
+    default = lib.maybeEnv "PASSWORD_STORE_DIR" ".password-store";
```
I would not rely on environment variables here. At least not at evaluation time. With flakes
we are trying very hard to get away from impure evaluation as it makes evaluation between different systems harder:

```suggestion
    default =  "/root/.password-store";
```

ju1m replied:
Thanks. I've fixed as you've suggested and just mentioned in the `description` that `getEnv PASSWORD_STORE_DIR` is a possible setting.

Mic92 commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,253 @@
+          # ssh root@example.org "unshare --mount sh -c '
+          #   mount --bind /var/empty /run/user;
+          #   gpg-preset-passphrase --homedir /var/lib/gnupg --preset $keygrip'"
+          ExecStartPre = "${pkgs.coreutils}/bin/install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts";
```
I think there might be a problem here: `/run/user/0` is a `tmpfs`. So its not just about deleting files but the whole filesystem is remounted. As a result this might lead to race conditions where a root user logins or out during deployment where the fs is mounted again.
If you really want to depend on an existing user-configured gpg-daemon (which I would not recommend because I don't see a sane way how this session can be assured to be present on boot) than the service should make its own login session. One way to do this would be a systemd user unit for root itself. The systemd user unit session could be started/stopped by this systemd service itself. It might be also possible to achieve the same using `systemd-run`.

ju1m replied:
Hmm, indeed it may, even if this `ExecStartPre=` is here only for the `gpg`/`gpg-preset-passphrase` calls to upload material, not for the `gpg-agent` (using `systemd.sockets.gpg-agent`) itself, nor for the `secret-*.service` (which redo the same in their loop) if no `secret-*.service` remain in the `activating` state, the `/run/user/0` could disappear for good if root logins/logouts, at this state it's ok because no `/run/user/0` also redirects to the right socket but when login again to upload the OpenPGP material, there would be a `/run/user/0` but no `/run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts` and the connection to the right `gpg-agent ` would fail.
The obvious solution I see to this is to require to run `install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts` in the upload command before calling `gpg`/`gpg-preset-passphrase`, it would be equivalent to setting a custom `GPG_AGENT_INFO` on old `gpg`.
Concerning the "user-configured gpg-daemon" you would not recommend, well, if you don't recommend it I'll stay with requiring this `install` on upload :\

Mic92 replied:
I think your install command can fail in the following case:

1. No tmpfs exists at /run/user/0, your service runs: `install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts`
2. A root user logins in and `/run/user/0` is mounted

I will probably still works because gpg starts a new daemon but now you have two gpg-daemons.
So the question is why not just set TMPDIR to /tmp and use PrivateTmp=true? Than gpg should use this directory but the daemon is private to your service and its socket not accessible to the outside.

Mic92 replied:
Ok. That does not work either because `/run` will be always used.

```
$ gpgconf --list-dirs
sysconfdir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/etc/gnupg
bindir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/bin
libexecdir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/libexec
libdir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/lib/gnupg
datadir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/share/gnupg
localedir:/nix/store/mi3axf21gf7lg8bilzxy8088ny4xwavg-gnupg-2.2.20/share/locale
socketdir:/run/user/1000/gnupg
dirmngr-socket:/run/user/1000/gnupg/S.dirmngr
agent-ssh-socket:/run/user/1000/gnupg/S.gpg-agent.ssh
agent-extra-socket:/run/user/1000/gnupg/S.gpg-agent.extra
agent-browser-socket:/run/user/1000/gnupg/S.gpg-agent.browser
agent-socket:/run/user/1000/gnupg/S.gpg-agent
homedir:/home/joerg/.gnupg
```

I am now actually in favor of:

```nix
BindPaths = ["/$sometmpfs:/run/user/0"];
```

This seems a lot less hacky.

Mic92 replied:
Because of gpg-daemon I tried to avoid all gpg shenanigans in `sops-nix`.

ju1m replied:
`TMPDIR` does not work because `/run/user` is hardcoded in gpg https://github.com/gpg/gnupg/blob/0904b8ef348a52335c378bee6dc90a978885d66f/common/homedir.c 
If the `install` is run just before the uploading `gpg` it's improbable that a root login could race it by mounting `/run/user/0`. 
Also,Maybe there is also a way to have this `install` run when root logins, by putting it in a `profile` or else.
Using `BindPaths=` would only work within the scope of the unit, and the uploading would have to use `unshare --mount sh -c 'mount --bind /$sometmpfs /run/user && gpg …'`, note that it would be `BindPaths = ["/$sometmpfs:/run/user"];` not `BindPaths = ["/$sometmpfs:/run/user/0"];` because `/run/user/0` must not exist for `gpg` to default to `${homedir}/S.gpg-agent`. Yes `gpg` is a pain on this point, or I'm trying to have a persistent gpg-daemon the wrong way and should indeed consider a user session, I don't know, anyway, I hope it goes smoother for you with `sops`.

Mic92 replied:
How about using `machinectl shell` to get a root logind session?

```
sudo machinectl shell .host $(which env) | grep RUNTIME
Connected to the local host. Press ^] three times within 1s to exit session.
XDG_RUNTIME_DIR=/run/user/0
```

Mic92 commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,253 @@
+          set -eux
+          decrypt() {
+            # In case /run/user/0 has been cleaned up by pam_systemd
+            install -D -m 640 -d /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
```
Again this seems racy. What uses case do you want to support beyond having gpg simply working?

ju1m replied:
This is here for the `gpg --decrypt` just after, in the case where the `install` in `ExecStartPre=` in `gpg-agent.service` has been wiped out by a root login/logout but root is logged in again so `/run/usr/0` is here again. The only race I see is between this install and the `gpg` right after, which seems highly improbable.
I thought it could also help for the uploading `gpg`/`gpg-preset-passphrase` commands but as you pointed out there was a problem, and now I require this `install` also before the uploading commands.

Mic92 replied:
Is the `BindPaths` above an appropriate replacement?

ju1m replied:
Hmm, strangely enough my comment about removing completely `/run/user/0` is wrong now, it can just be empty, but I could have sworn emptying `/run/user/0` was not enough in my tests 12h ago. *shrugs*
```
$ ssh root@localhost
# gpg -K
# file /run/user/0/gnupg/S.gpg-agent
/run/user/0/gnupg/S.gpg-agent: socket
```
Both `unshare` and `system-run` work fine.
```
# unshare --mount sh -c "mkdir -p /run/user/0/gnupg && mount --bind /var/empty /run/user/0/gnupg && gpgconf --homedir $PWD --list-dirs agent-socket"
/var/lib/gnupg/S.gpg-agent
```
```
# systemd-run -P -p BindPaths='/var/empty:/run/user/0/gnupg' gpgconf --homedir /var/lib/gnupg --list-dirs agent-socket
Running as unit: run-u4801.service
/var/lib/gnupg/S.gpg-agent
```
I'll change the `install` hack to that `BindPaths=` hack, making it `BindReadOnlyPaths=`, in the services and in the uploading through `systemd-run -P`.

Mic92 replied:
Ok. The other alternative would be to get a proper logind session with machinectl as noted above.

ju1m replied:
Hmm, this works also: `InaccessiblePaths=/run/user/0` but not: `InaccessiblePaths=/run/user/0/gnupg`. Strange.

ju1m replied:
Oh, ok, `InaccessiblePaths=/run/user/0` does work either  if `/run/user/0` does not exist. So AFAICS that could only be `InaccessiblePaths=/run/user` or `BindReadOnlyPaths=/var/empty:/run/user/0/gnupg`

ju1m replied:
Good news! In fact `pam_systemd` uses a unit to setup `/run/user/0` and an `ExecStartPost=` running the `install` hack can be added to the instance of the template used, by writing to `/etc/systemd/system/user-runtime-dir@0.service.d/override.conf` (which requires another hack in NixOS, but is doable). Also, because `ExecStartPost=` is taken into account for the purpose of `Before=`/`After=` ordering constraints, this prevents race condition regarding the uploading commands. And for the decrypting services I've kept the `BindReadOnlyPaths=` hack.

Mic92 commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,271 @@
+    lib.mapAttrs' (target: secret:
+      lib.nameValuePair (lib.removeSuffix ".service" secret.service) {
+        description = "Install secret ${secret.path}";
+        after = ["gpg-agent.service"];
```
So does this service starts its own gpg daemon? Because it mounts `/var/lib/gnupg/empty` ?

Mic92 commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,271 @@
+      };
+    } //
+    lib.mapAttrs' (target: secret:
+      lib.nameValuePair (lib.removeSuffix ".service" secret.service) {
```
So does this service starts a new gnupg daemon? Because it bind mounts `/var/lib/gnupg/empty`?

bqv commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,271 @@
+          set -o pipefail
+          set -eux
+          decrypt() {
+            ${pkgs.gnupg}/bin/gpg --homedir '${gnupgHome}' --no-autostart --batch --decrypt '${secret.gpg}' |
```
Is there a newline escape missing here?

8573 replied:
If you mean to continue the command across the line-break, the pipe makes it unnecessary:

```
$ echo foo |
> cat
foo
```

(This works at least in Bash, dash, and Zsh.)

bqv replied:
Interesting. ignore this then!

d-goldin commented:
Sorry for the late response, was kinda swamped. 

Removing myself from reviewers, since I'm personally not very keen on an approach tied to a specific "backend", but also don't want to block this PR in any way.

Aside from that, without having tried it, I think this is a rather small addition easy enough to get in and can be useful enough to some people/setups with only moderate transition costs should something else come around.

What I think could be important here to get more feedback is an updated documentation on the usage scenarios early on, including scenarios in which this would not work well (i.e. to aid reviewers/testers gain understanding quicker). For instance, from first glance it seems to me like this would not work with services utilizing `DynamicUser`, of which we can expect more in the future based on https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md. Maybe it would be possible to add some checks to break early on for services which don't fulfill those potential expectations.

What would be also good to check is whether this could benefit from some likely upcoming related changes in systemd and how this might change the interface (see https://github.com/systemd/systemd/pull/16568).

Mic92 commented:
> Sorry for the late response, was kinda swamped.
> 
> Removing myself from reviewers, since I'm personally not very keen on an approach tied to a specific "backend", but also don't want to block this PR in any way.
> 
> Aside from that, without having tried it, I think this is a rather small addition easy enough to get in and can be useful enough to some people/setups with only moderate transition costs should something else come around.
> 
> What I think could be important here to get more feedback is an updated documentation on the usage scenarios early on, including scenarios in which this would not work well (i.e. to aid reviewers/testers gain understanding quicker). For instance, from first glance it seems to me like this would not work with services utilizing `DynamicUser`, of which we can expect more in the future based on [https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md](https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md?rgh-link-date=2020-08-18T19%3A01%3A55Z). Maybe it would be possible to add some checks to break early on for services which don't fulfill those potential expectations.
> 
> What would be also good to check is whether this could benefit from some likely upcoming related changes in systemd and how this might change the interface (see [systemd/systemd#16568](https://github.com/systemd/systemd/pull/16568)).

DynamicUser can be also supported like this: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/monitoring/prometheus/alertmanager.nix#L170
By using envsubst and EnvironmentFile one can work-around permission.

bqv commented on pkgs/tools/security/gnupg/22.nix:
```diff
@@ -69,6 +69,9 @@ stdenv.mkDerivation rec {
     ln -s $out/bin/gpg $out/bin/gpg2
+
+    # Make libexec tools available in PATH
+    ln -s -t $out/bin $out/libexec/*
```
You could separate this change out into it's own PR, to give this a better chance of being merged, since I'm assuming this is what is causing the hundreds of rebuilds

ju1m replied:
Done in https://github.com/NixOS/nixpkgs/pull/100031, thanks.

ju1m committed:
nixos/security.gnupg: init

bqv commented:
With a bump, this is now merely a new module!

bqv commented:
d5d543d2636 is in :) with a bump, this is now merely a new module!

bqv commented:
@Mic92 merge?

ju1m commented:
FYI, I've updated the first comment to put an example `flake.nix` which provides a `nix run .#$MACHINE.sendkeys`. It's a stripped version of a `flake.nix` I use, I hope I haven't made any mistake in the stripping process.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/385

ju1m commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,266 @@
+          default = "secret-" + escapeUnitName name + ".service";
+          description = ''
+            The name of the systemd service.
+            Useful to put constraints like <literal>after</literal> or <literal>wants</wants>
```
@bqv, thanks

ju1m commented:
Added `restartIfChanged = false` to keep `gpg-agent` alive to preserve its cache, which avoids `switch-to-configuration switch` to wait for passwords to be given again.

seylerius commented:
I think this is a really great idea, and I'm looking forward to the eventual merge. Hopefully it should be accessible from Home-Manager as well, which will allow me to separate all my API keys and tokens and other secrets entirely out of my configs, and make it easier for me to keep the lot of it in a git repo for easier fetching on new machines. Great work here.

rycee commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,270 @@
+      options = {
+        gpg = lib.mkOption {
+          type = types.path;
+          default =
```
Perhaps this will cause the manual to depend on the configuration? It might be necessary to use `defaultText` to break this dependency.

rycee commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,270 @@
+      <literal>builtins.getEnv "PASSWORD_STORE_DIR" + "/machines/example"</literal>.
+    '';
+  };
+  secrets = lib.mkOption {
```
I typically find it improves readability and navigability if there is an empty line between option definitions.

rycee commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,270 @@
+        };
+        service = lib.mkOption {
+          type = types.str;
+          default = "secret-" + escapeUnitName name + ".service";
```
Also causes manual build dependency?

rycee commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,270 @@
+        postStop = lib.mkOption {
+          type = types.lines;
+          default = "";
+          example = ''shred -u "$secret_file"'';
```
I think this may be a dangerous suggestion since people may uncritically copy this command to their own configuration without carefully considering whether it is effective on their file system.

ju1m replied:
Thanks. I've moved the `postStart` and `postStop` options into `systemdConfig` which is also useful to set ` before`, `wantedBy` or `requiredBy` directly when configuring the secret instead of using `systemd.services.foo.after = [ secrets."path/to/secret".service ];`

rycee commented on nixos/modules/security/gnupg.nix:
```diff
@@ -0,0 +1,270 @@
+config = lib.mkIf (cfg.secrets != {}) {
+  # Because /run/user/0 is wiped out by pam_systemd when root logouts,
+  # systemd.services.gpg-agent cannot put its socket in
+  # the path expected by gpg: /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
```
Maybe I'm missing something obvious but why the `d.6qoenf9br6fajbkknuz1i6ts` name?

ju1m replied:
It's the standard socket path used by GnuPG, derived from the given `--homedir`. With the workarounds used here, `gpg`/`gpg-agent` cannot create this standard socket and default to `<homedir>/S.gpg-agent`. This is wanted here to keep the `gpg-agent` alive, because when `root` logouts its `/run/user/0/` is removed and `gpg-agent` would no longer be accessible, detect that and kill itself.
See also https://github.com/gpg/gnupg/commit/aab8a0b05292b0d06e3001a0b289224cb7156dbd

ju1m commented:
I've updated according to @rycee's review. Beware that I've also removed `wantedBy = ["multi-user.target"];` for the decrypting services, to avoid decrypting unused secrets, this means that you now **must** add `systemd` dependencies indications (`after`/`before` and `required`/`requiredBy` or `wants`/`wantedBy`+`postStart = "systemctl reload foo"`), example:
```
security.gnupg.secrets."tor/onion/${onion}/hs_ed25519_secret_key" = {
  systemdConfig.before = [ "tor.service" ];
  systemdConfig.requiredBy = [ "tor.service" ];
};
```
I've also moved `postStart` and `postStop` to `systemdConfig.postStart` and `systemdConfig.postStop`.

ju1m commented:
Introducing the possiblity to `nix -L run .#nixosConfigurations.${hostName}.config.security.gnupg.agent.sendKeys`.
I've also changed `--homedir` from `/var/lib/gnupg` to `/var/lib/gnupg/.gnupg` to be able to `ssh gnupg@` without setting a `GNUPGHOME=` (though I could). At this point, I'm not making `ssh gnupg@` the default because using `root` is more straightforward for the "sending and installing secrets" use case. But the `gnupg` user and the `gnupg` group are now there to explore more use cases.

ju1m commented:
Opt-out `security.gnupg.agent.enable` by default because it can now be enabled without any secret configured. Useful to start `gpg-agent.service` in a first rebuild without blocking indefinitely when sending secret that can't be installed because the password could not have been sent before that first rebuild activating `gpg-agent.service`, as explained in the updated intro.

ju1m commented:
Improved the documentation a little bit and fixed `UMask` typo.

ju1m commented:
Fixed permission of `/run/user/$uid` for non-`root` users.

ju1m commented:
Sorry, I've forgotten I had rebased a bit so Github's [force-pushed link](https://github.com/NixOS/nixpkgs/compare/797680ac249e53938e2d6cbcefdae3535b88c075..881a0aecbfdff77278f1b3d511ea475eef5afbb4) is not pretty; the diff is b3dc3af437f..58410e3735e, changelog is:
- [X] Support relative paths in `security.gnupg.agent.keyring.${keygrip}.passwordGpg`, like "gnupg/root.gpg". I am keeping the default to "keygrip/${keygrip}.gpg", but I know it's a bit catch22 when the key is created from an encrypted password which can't obviously already have the keygrip in it's path.
- [X] Add `"-o" "StrictHostKeyChecking=yes"` to the default `security.gnupg.agent.keyring.${keygrip}.ssh`, to better ensure secrets are not sent where they should not.
- [X] Use `lib.mapNullable`.
- [X] Fix typo in tracing comment.
